## **Finance Database**
-------

##### **Deskripsi Program**
Tujuan dari program ini adalah untuk mengelola data dari divisi Finance yang sebelummnya tersedia dalam format CSV. Proses ini mencakup *cleaning* dan normalisasi data, pembuatan database di PostgreSQL, penentuan akses query, dan pengujian keberhasilan database tersebut.

#### **1. Data Exploration** 

##### a. Load file data dengan Pandas lalu tampilkan data. Informasi apa yang bisa kamu sampaikan dari langkah ini?

In [32]:
# Import library pandas
import pandas as pd

# Melakukan load dan read file csv dari URL dan menyimpannya ke dalam variabel
df = pd.read_csv('https://raw.githubusercontent.com/FTDS-learning-materials/phase-0/main/src/Financials.csv')

# Menampilkan data yang tersimpan
df

Segment                   Country     Product   Discount Band   \
0          Government                    Canada   Carretera            None    
1          Government                   Germany   Carretera            None    
2           Midmarket                    France   Carretera            None    
3           Midmarket                   Germany   Carretera            None    
4           Midmarket                    Mexico   Carretera            None    
..                ...                       ...          ...             ...   
695    Small Business                    France    Amarilla            High    
696    Small Business                    Mexico    Amarilla            High    
697        Government                    Mexico     Montana            High    
698        Government                    Canada       Paseo            High    
699  Channel Partners  United States of America         VTT            High    

     Units Sold   Manufacturing Price   Sale Price     Gross Sales   \
0     $1,618.50                 $3.00       $20.00      $32,370.00    
1     $1,321.00                 $3.00       $20.00      $26,420.00    
2     $2,178.00                 $3.00       $15.00      $32,670.00    
3       $888.00                 $3.00       $15.00      $13,320.00    
4     $2,470.00                 $3.00       $15.00      $37,050.00    
..           ...                   ...          ...             ...   
695   $2,475.00               $260.00      $300.00    $7,42,500.00    
696     $546.00               $260.00      $300.00    $1,63,800.00    
697   $1,368.00                 $5.00        $7.00       $9,576.00    
698     $723.00                $10.00        $7.00       $5,061.00    
699   $1,806.00               $250.00       $12.00      $21,672.00    

         Discounts           Sales            COGS        Profit         Date  \
0             $-        $32,370.00      $16,185.00    $16,185.00   01/01/2014   
1             $-        $26,420.00      $13,210.00    $13,210.00   01/01/2014   
2             $-        $32,670.00      $21,780.00    $10,890.00   01/06/2014   
3             $-        $13,320.00       $8,880.00     $4,440.00   01/06/2014   
4             $-        $37,050.00      $24,700.00    $12,350.00   01/06/2014   
..              ...             ...             ...           ...         ...   
695   $1,11,375.00    $6,31,125.00    $6,18,750.00    $12,375.00   01/03/2014   
696     $24,570.00    $1,39,230.00    $1,36,500.00     $2,730.00   01/10/2014   
697      $1,436.40       $8,139.60       $6,840.00     $1,299.60   01/02/2014   
698        $759.15       $4,301.85       $3,615.00       $686.85   01/04/2014   
699      $3,250.80      $18,421.20       $5,418.00    $13,003.20   01/05/2014   

     Month Number  Month Name   Year  
0               1     January   2014  
1               1     January   2014  
2               6        June   2014  
3               6        June   2014  
4               6        June   2014  
..            ...          ...   ...  
695             3       March   2014  
696            10     October   2014  
697             2    February   2014  
698             4       April   2014  
699             5         May   2014  

[700 rows x 16 columns]

Pertama-tama, untuk mengakses *library* yang ada di Pandas, perlu di-import *library*nya terlebih dahulu. Setelah itu, untuk meload dan membaca file dapat menggunakan metode `read_csv`.

Berdasarkan data di atas, dapat dilihat bahwa dataframe merupakan laporan keuangan yang berisi penjualan produk berdasarkan pendapatan, profit dari produk, diskon, dan informasi lainnya dari masing-masing negara. Dataframe ini memiliki 700 baris dan 16 kolom. Kolom terdiri dari `Segment`, `Country`, `Product`, `Discount Band`, `Units Sold`, `Manufacturing Price`, `Sale Price`, `Gross Sales`, `Discounts`, `Sales`, `COGS`, `Profit`, `Date`, `Month Number`, `Month Name`, dan `Year`. Dengan melakukan langkah ini, terdapat gambaran awal mengenai struktur dan isi data yang tersedia dalam dataframe tersebut.

##### b. Tampilkan informasi *summary data* dan berikan penjelasan mengenai temuan yang diperoleh.

In [33]:
# Menampilkan ringkasan informasi mengenai dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Segment                700 non-null    object
 1   Country                700 non-null    object
 2    Product               700 non-null    object
 3    Discount Band         700 non-null    object
 4    Units Sold            700 non-null    object
 5    Manufacturing Price   700 non-null    object
 6    Sale Price            700 non-null    object
 7    Gross Sales           700 non-null    object
 8    Discounts             700 non-null    object
 9     Sales                700 non-null    object
 10   COGS                  700 non-null    object
 11   Profit                700 non-null    object
 12  Date                   700 non-null    object
 13  Month Number           700 non-null    int64 
 14   Month Name            700 non-null    object
 15  Year                   

Untuk menampilkan informasi *summary data* dapat menggunakan metode info. Berdasarkan informasi *summary data*, dataset ini memiliki 700 baris dan 16 kolom. Melalui metode ini, dapat dilihat bahwa setiap kolom memiliki 700 *non-null values*, menunjukkan bahwa pada saat ini, dataset tidak mengandung nilai yang hilang *(missing values)*. Dari total 16 kolom, dua di antaranya memiliki tipe data `integer`, sementara 14 kolom lainnya memiliki tipe data `object`.

In [34]:
df.describe()

Month Number         Year
count    700.000000   700.000000
mean       7.900000  2013.750000
std        3.377321     0.433322
min        1.000000  2013.000000
25%        5.750000  2013.750000
50%        9.000000  2014.000000
75%       10.250000  2014.000000
max       12.000000  2014.000000

Untuk mengetahui informasi statistik dari data numerik dalam DataFrame, dapat menggunakan metode `describe`. Metode di atas menunjukkan nilai-nilai statistik dari masing-masing kolom numerik untuk mendapatkan gambaran umum mengenai distribusi datanya. Akan tetapi, dikarenakan hanya terdapat 2 kolom yang memiliki tipe data `integer`, maka yang ditampilkan hanya 2 kolom tersebut. Kolom lainnya yang memiliki angka tetapi memiliki tipe data `object` tidak dapat terhitung.

In [35]:
df.value_counts()

Segment           Country                    Product    Discount Band    Units Sold    Manufacturing Price    Sale Price    Gross Sales    Discounts       Sales         COGS           Profit        Date        Month Number   Month Name   Year
Channel Partners  Canada                    Amarilla   High             $2,761.00     $260.00                $12.00        $33,132.00     $3,975.84      $29,156.16     $8,283.00      $20,873.16     01/09/2013  9             September     2013    1
Government        United States of America  Paseo      Low              $274.00       $10.00                 $350.00       $95,900.00     $3,836.00      $92,064.00     $71,240.00     $20,824.00     01/12/2014  12            December      2014    1
                                                       High             $1,177.00     $10.00                 $350.00       $4,11,950.00   $57,673.00     $3,54,277.00   $3,06,020.00   $48,257.00     01/11/2014  11            November      2014    1
             

Metode `value_counts` digunakan untuk menghitung jumlah nilai unik yang ada dalam sebuah Series atau kolom DataFrame. Hal ini memberikan informasi tentang frekuensi munculnya nilai-nilai tersebut dalam data.

Dapat dilihat bahwa rata-rata frekuensi yang muncul hanya 1 kali dan dari *length* menunjukkan angka 700 yang berarti tidak terdapat duplikat atau nilai yang muncul lebih dari sekali pada dataframe ini. Hal ini menunjukkan bahwa seluruh nilai pada dataframe sudah bersifat unik.

Oleh karena itu, disimpulkan bahwa ketika melalukan data *cleaning*, tidak perlu melakukan penghapusan duplikat karena tidak ada data duplikat pada dataframe. 

In [36]:
df.dtypes

Segment                  object
Country                  object
 Product                 object
 Discount Band           object
 Units Sold              object
 Manufacturing Price     object
 Sale Price              object
 Gross Sales             object
 Discounts               object
  Sales                  object
 COGS                    object
 Profit                  object
Date                     object
Month Number              int64
 Month Name              object
Year                      int64
dtype: object

Metode `dtypes` dapat menampilkan tipe data dari masing-masing kolom pada dataframe. Dari informasi yang telah disampaikan, terdapat beberapa kolom dengan tipe data yang tidak sesuai dengan nilai-nilai yang terdapat di dalamnya. Untuk mempermudah analisis data ke depan, diperlukan perubahan tipe data kolom tersebut agar sesuai dengan nilai-nilai yang sebenarnya.

Beberapa kolom yang berisi angka memiliki karakter-karakter yang membuat data type nya menjadi object. Data type tersebut perlu diubah tipe datanya menjadi float, integer, maupun datetime sesuai dengan jenis data yang ada di dalamnya. Hal ini akan membantu dalam melakukan analisis lebih lanjut terhadap data tersebut.

##### c. Tampilkan list nama-nama kolom dan berikan penjelasan temuan apa yang kamu dapatkan, serta berikan langkah apa selanjutnya dari temuan ini.

In [37]:
# Menampilkan list nama-nama kolom
df.columns

Index(['Segment', 'Country', ' Product ', ' Discount Band ', ' Units Sold ',
       ' Manufacturing Price ', ' Sale Price ', ' Gross Sales ', ' Discounts ',
       '  Sales ', ' COGS ', ' Profit ', 'Date', 'Month Number',
       ' Month Name ', 'Year'],
      dtype='object')

Metode `columns` dapat menampilkan list nama-nama kolom dari dataframe. Dapat dilihat bahwa penamaan dari beberapa kolom masih ada yang kurang tepat, yaitu terdapat *whitespace* yang tidak diperlukan. Oleh karena itu, langkah berikutnya yang harus dilakukan adalah mengubah penamaan kolom dengan format yang lebih rapih dengan menghilangkan *whitespace* yang tidak diperlukan.

##### d. Tampilkan list nilai unik masing-masing kolom `Segment`, `Country`, `Product`, `Discount Band` dan berikan informasi apa yang dapat kamu ambil dari langkah ini.

In [38]:
# Menampilkan list nilai unik dari kolom Segment
df['Segment'].unique()

array(['Government', 'Midmarket', 'Channel Partners', 'Enterprise',
       'Small Business'], dtype=object)

In [39]:
# Menampilkan list nilai unik dari kolom Country
df['Country'].unique()

array(['Canada', 'Germany', 'France', 'Mexico',
       'United States of America'], dtype=object)

In [40]:
# Menampilkan list nilai unik dari kolom Product
df[' Product '].unique()

array([' Carretera ', ' Montana ', ' Paseo ', ' Velo ', ' VTT ',
       ' Amarilla '], dtype=object)

In [41]:
# Menampilkan list nilai unik dari kolom Discount Band
df[' Discount Band '].unique()

array([' None ', ' Low ', ' Medium ', ' High '], dtype=object)

Untuk menampilkan list nilai unik dari masing-masing kolom dapat menggunakan metode `unique`. Dari seluruh *unique values* yang didapatkan di atas, terlihat bahwa terdapat kesalahan penamaan isi list pada kolom `Product` dan `Discount Band`. Penamaan list tersebut mengandung *whitespace* yang seharusnya tidak ada sehingga perlu diperbaiki dengan menghapus *whitespace* yang tidak sesuai.

#### **2. Data Cleaning**

> Setelah melakukan eksplorasi data sebelumnya, terdapat beberapa langkah yang dapat dilakukan dalam proses pembersihan data *(data cleaning)*:
> - Perbaikan penamaan kolom yang mengandung *whitespace* yang tidak perlu.
> - Perbaikan penamaan list dari isi kolom yang mengandung *whitespace* yang tidak perlu.
> - Penghapusan karakter-karakter yang tidak perlu.
> - Penyesuaian tipe data pada beberapa kolom agar sesuai dengan jenis data yang sebenarnya.

##### a. Manipulasi kolom meliputi penggantian nama kolom, penghapusan kolom, atau penambahan kolom

In [42]:
# Mengganti penamaan kolom yang mengandung spasi yang tidak perlu

df = df.rename(columns={' Product ': 'Product',
                        ' Discount Band ': 'Discount Band',
                        ' Units Sold ': 'Units Sold',
                        ' Manufacturing Price ': 'Manufacturing Price',
                        ' Sale Price ': 'Sale Price',
                        ' Gross Sales ': 'Gross Sales',
                        ' Discounts ': 'Discounts',
                        '  Sales ': 'Sales',
                        ' COGS ': 'COGS',
                        ' Profit ': 'Profit',
                        ' Month Name ': 'Month Name',
                        })

# Menampilkan nama kolom setelah perubahan
df.columns

Index(['Segment', 'Country', 'Product', 'Discount Band', 'Units Sold',
       'Manufacturing Price', 'Sale Price', 'Gross Sales', 'Discounts',
       'Sales', 'COGS', 'Profit', 'Date', 'Month Number', 'Month Name',
       'Year'],
      dtype='object')

Perbaikan nama kolom dapat dilakukan dengan metode `rename`. Dalam tahap ini, dilakukan perbaikan penamaan kolom yang sebelumnya mengandung *whitepace* yang tidak diperlukan. Setelah dilakukan perubahan, nama-nama kolom akan terlihat lebih rapi dan sesuai dengan standar penamaan yang lebih baik.

Pada tahap ini,  tidak diperlukan penghapusan kolom karena tidak terdapat data yang duplikat yang perlu dihilangkan. Selain itu, juga tidak perlu dilakukan penambahan kolom karena tidak ada informasi tambahan yang perlu ditambahkan ke dalam dataset. Tahap ini fokus pada perbaikan penamaan kolom.

##### b. Penggantian tipe data yang sesuai dengan seharusnya. Untuk nilai harga dengan simbol mata uang, simbol tersebut dapat dibuang supaya dapat terkonversi ke tipe data yang seharusnya. Perhatikan format penulisan angkanya apakah ada tanda yang membuat angka hanya bisa bertipe data string.
> (Tambahan: Di suatu kolom yang bernilai uang, terpadat beberapa data yang mengandung karakter '(' dan ')' di dalamnya sehingga harus ditake-out karakter tersebut.)

##### c. Cek dan handling missing value. (Hint: terdapat non-standard missing value pada kolom yang semestinya bertipe data integer, handling dapat dilakukan dengan imputasi dengan angka nol '0')

In [43]:
# Menghapus spasi yang tidak diperlukan pada isi list kolom 

df['Product'] = df['Product'].str.replace(' ', '')
df['Discount Band'] = df['Discount Band'].str.replace(' ', '')

Dari analisis sebelumnya, didapatkan bahwa dalam kolom `Product` dan `Discount Band`, terdapat spasi yang tidak diperlukan pada nama item di dalam list. Untuk memperbaiki penamaan dari isi list tersebut, dapat menggunakan metode `str.replace`. Hal ini dapat membantu memperbaiki penamaan sehingga tidak terdapat spasi yang tidak diperlukan.

In [44]:
# Menyimpan nama-nama kolom pada sebuah variabel
cleanColumns = ['Units Sold', 'Manufacturing Price', 'Sale Price', 'Gross Sales', 'Discounts', 'Sales', 'COGS', 'Profit']

# Melakukan looping pada cleanColumns untuk mengubah data
for col in cleanColumns:
    # Menghapus seluruh karakter maupun spasi selain digit dan desimal
    df[col] = df[col].str.replace(r'[^\d.]+', '', regex=True)
    # Mengisi data yang kosong (missing values) menjadi angka 0
    df[col] = df[col].replace('', '0')

Untuk mengubah tipe data ke format yang sesuai, langkah pertama adalah membersihkan karakter-karakter yang tidak diperlukan, seperti simbol uang ($), koma (,), tanda strip (-), dan lainnya. Langkah ini dilakukan dengan menggunakan metode `str.replace`.

Setelah membersihkan karakter-karakter tersebut, beberapa kolom dapat menjadi kosong *(null)* karena sebelumnya hanya berisi karakter. Untuk menghindari nilai null, langkah selanjutnya adalah mengganti nilai-nilai kosong tersebut dengan angka '0'.

In [45]:
# Mengubah tipe data menjadi datetime

df['Date'] = pd.to_datetime(df['Date'])
# df['Month Number'] = pd.to_datetime(df['Month Number'], format='%m').dt.month

Setelah membersihkan karakter dan menghindari nilai null, selanjutnya adalah mengubah tipe data `Date` menjadi datetime dengan metode `pd.to_datetime`.

In [46]:
# Mengubah tipe data menjadi float

df['Units Sold'] = df['Units Sold'].astype(float)
df['Manufacturing Price'] = df['Manufacturing Price'].astype(float)
df['Sale Price'] = df['Sale Price'].astype(float)
df['Gross Sales'] = df['Gross Sales'].astype(float)
df['Discounts'] = df['Discounts'].astype(float)
df['Sales'] = df['Sales'].astype(float)
df['COGS'] = df['COGS'].astype(float)
df['Profit'] = df['Profit'].astype(float)

Setelah itu, data yang berisi angka diubah dari `Object` menjadi `Float` dengan metode `astype`.

In [47]:
# Menampilkan kembali summary data setelah perubahan
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Segment              700 non-null    object        
 1   Country              700 non-null    object        
 2   Product              700 non-null    object        
 3   Discount Band        700 non-null    object        
 4   Units Sold           700 non-null    float64       
 5   Manufacturing Price  700 non-null    float64       
 6   Sale Price           700 non-null    float64       
 7   Gross Sales          700 non-null    float64       
 8   Discounts            700 non-null    float64       
 9   Sales                700 non-null    float64       
 10  COGS                 700 non-null    float64       
 11  Profit               700 non-null    float64       
 12  Date                 700 non-null    datetime64[ns]
 13  Month Number         700 non-null  

Dapat dilihat bahwa tipe data dari masing-masing kolom sudah sesuai dengan jenisnya masing-masing. Dari seluruh data, terdapat 1 tipe data `Datetime`,  8 tipe data `Float`, 2 tipe data `Integer` , dan 5 tipe data `Object`

In [48]:
# Menampilkan kembali dataframe setelah perubahan
df

Segment                   Country    Product Discount Band  \
0          Government                    Canada  Carretera          None   
1          Government                   Germany  Carretera          None   
2           Midmarket                    France  Carretera          None   
3           Midmarket                   Germany  Carretera          None   
4           Midmarket                    Mexico  Carretera          None   
..                ...                       ...        ...           ...   
695    Small Business                    France   Amarilla          High   
696    Small Business                    Mexico   Amarilla          High   
697        Government                    Mexico    Montana          High   
698        Government                    Canada      Paseo          High   
699  Channel Partners  United States of America        VTT          High   

     Units Sold  Manufacturing Price  Sale Price  Gross Sales  Discounts  \
0        1618.5                  3.0        20.0      32370.0       0.00   
1        1321.0                  3.0        20.0      26420.0       0.00   
2        2178.0                  3.0        15.0      32670.0       0.00   
3         888.0                  3.0        15.0      13320.0       0.00   
4        2470.0                  3.0        15.0      37050.0       0.00   
..          ...                  ...         ...          ...        ...   
695      2475.0                260.0       300.0     742500.0  111375.00   
696       546.0                260.0       300.0     163800.0   24570.00   
697      1368.0                  5.0         7.0       9576.0    1436.40   
698       723.0                 10.0         7.0       5061.0     759.15   
699      1806.0                250.0        12.0      21672.0    3250.80   

         Sales      COGS    Profit       Date  Month Number  Month Name  Year  
0     32370.00   16185.0  16185.00 2014-01-01             1    January   2014  
1     26420.00   13210.0  13210.00 2014-01-01             1    January   2014  
2     32670.00   21780.0  10890.00 2014-01-06             6       June   2014  
3     13320.00    8880.0   4440.00 2014-01-06             6       June   2014  
4     37050.00   24700.0  12350.00 2014-01-06             6       June   2014  
..         ...       ...       ...        ...           ...         ...   ...  
695  631125.00  618750.0  12375.00 2014-01-03             3      March   2014  
696  139230.00  136500.0   2730.00 2014-01-10            10    October   2014  
697    8139.60    6840.0   1299.60 2014-01-02             2   February   2014  
698    4301.85    3615.0    686.85 2014-01-04             4      April   2014  
699   18421.20    5418.0  13003.20 2014-01-05             5        May   2014  

[700 rows x 16 columns]

Berikut tampilan akhir dari data yang sudah mengalami *cleaning*:
- Penamaan kolom dan isi dari kolom sudah sesuai dengan kaidah penamaan yang benar.
- Karakter dan spasi yang tidak diperlukan telah dihapus dari kolom-kolom yang membutuhkan.
- *Missing values* (nilai yang hilang) telah diubah menjadi '0' agar data lebih lengkap dan konsisten.
- Tipe data pada setiap kolom sudah sesuai dengan format yang diharapkan, sehingga memudahkan dalam analisis data lebih lanjut.

#### **3. Data Normalization**

##### a. Lakukan normalisasi data yang sudah diolah sebelumnya sampai dengan 3NF.

##### b. Jumlah tabel setelah dinormalisasi sebanyak 5 tabel termasuk tabel utama.

##### c. Perhatikan penggunaan primary key dan foreign key antar tabel.

Dalam melakukan normalisasi database, langkah pertama yang dilakukan adalah memastikan bahwa data sudah memenuhi aturan 1NF *(First Normal Form)*, dimana setiap kolom hanya berisi satu nilai. Setelah memastikan hal tersebut, langkah selanjutnya adalah melakukan normalisasi ke tingkat 2NF *(Second Normal Form)* dan 3NF *(Third Normal Form)*.

Dari data yang telah diperiksa, dapat disimpulkan bahwa dataframe sudah memenuhi aturan 1NF karena setiap kolom hanya berisi satu nilai. Oleh karena itu, langkah selanjutnya adalah melakukan normalisasi ke tingkat 2NF dan 3NF.

In [49]:
# Membuat tabel segment yang berisi isi list dan menyimpannya ke variabel

segments = pd.DataFrame(df['Segment'].unique())
segments

0
0        Government
1         Midmarket
2  Channel Partners
3        Enterprise
4    Small Business

In [50]:
# Mengganti nama isi segment menjadi index

i = 0 

# Melakukan proses looping untuk mengakses seluruh baris
for mainName in df['Segment']:
    for listName in segments.index:
        if mainName == segments.iloc[listName][0]:
            df.loc[i, 'Segment'] = listName + 1
    i += 1

df

Segment                   Country    Product Discount Band  Units Sold  \
0         1                    Canada  Carretera          None      1618.5   
1         1                   Germany  Carretera          None      1321.0   
2         2                    France  Carretera          None      2178.0   
3         2                   Germany  Carretera          None       888.0   
4         2                    Mexico  Carretera          None      2470.0   
..      ...                       ...        ...           ...         ...   
695       5                    France   Amarilla          High      2475.0   
696       5                    Mexico   Amarilla          High       546.0   
697       1                    Mexico    Montana          High      1368.0   
698       1                    Canada      Paseo          High       723.0   
699       3  United States of America        VTT          High      1806.0   

     Manufacturing Price  Sale Price  Gross Sales  Discounts      Sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         COGS    Profit       Date  Month Number  Month Name  Year  
0     16185.0  16185.00 2014-01-01             1    January   2014  
1     13210.0  13210.00 2014-01-01             1    January   2014  
2     21780.0  10890.00 2014-01-06             6       June   2014  
3      8880.0   4440.00 2014-01-06             6       June   2014  
4     24700.0  12350.00 2014-01-06             6       June   2014  
..        ...       ...        ...           ...         ...   ...  
695  618750.0  12375.00 2014-01-03             3      March   2014  
696  136500.0   2730.00 2014-01-10            10    October   2014  
697    6840.0   1299.60 2014-01-02             2   February   2014  
698    3615.0    686.85 2014-01-04             4      April   2014  
699    5418.0  13003.20 2014-01-05             5        May   2014  

[700 rows x 16 columns]

In [51]:
# Membuat tabel countries yang berisi isi list dan menyimpannya ke variabel

countries = pd.DataFrame(df['Country'].unique())
countries

0
0                    Canada
1                   Germany
2                    France
3                    Mexico
4  United States of America

In [52]:
# Mengganti nama isi country menjadi index

i = 0

# Melakukan proses looping untuk mengakses seluruh baris
for mainName in df['Country']:
    for listName in countries.index:
        if mainName == countries.iloc[listName][0]:
            df.loc[i, 'Country'] = listName + 1
    i += 1

df

Segment Country    Product Discount Band  Units Sold  Manufacturing Price  \
0         1       1  Carretera          None      1618.5                  3.0   
1         1       2  Carretera          None      1321.0                  3.0   
2         2       3  Carretera          None      2178.0                  3.0   
3         2       2  Carretera          None       888.0                  3.0   
4         2       4  Carretera          None      2470.0                  3.0   
..      ...     ...        ...           ...         ...                  ...   
695       5       3   Amarilla          High      2475.0                260.0   
696       5       4   Amarilla          High       546.0                260.0   
697       1       4    Montana          High      1368.0                  5.0   
698       1       1      Paseo          High       723.0                 10.0   
699       3       5        VTT          High      1806.0                250.0   

     Sale Price  Gross Sales  Discounts      Sales      COGS    Profit  \
0          20.0      32370.0       0.00   32370.00   16185.0  16185.00   
1          20.0      26420.0       0.00   26420.00   13210.0  13210.00   
2          15.0      32670.0       0.00   32670.00   21780.0  10890.00   
3          15.0      13320.0       0.00   13320.00    8880.0   4440.00   
4          15.0      37050.0       0.00   37050.00   24700.0  12350.00   
..          ...          ...        ...        ...       ...       ...   
695       300.0     742500.0  111375.00  631125.00  618750.0  12375.00   
696       300.0     163800.0   24570.00  139230.00  136500.0   2730.00   
697         7.0       9576.0    1436.40    8139.60    6840.0   1299.60   
698         7.0       5061.0     759.15    4301.85    3615.0    686.85   
699        12.0      21672.0    3250.80   18421.20    5418.0  13003.20   

          Date  Month Number  Month Name  Year  
0   2014-01-01             1    January   2014  
1   2014-01-01             1    January   2014  
2   2014-01-06             6       June   2014  
3   2014-01-06             6       June   2014  
4   2014-01-06             6       June   2014  
..         ...           ...         ...   ...  
695 2014-01-03             3      March   2014  
696 2014-01-10            10    October   2014  
697 2014-01-02             2   February   2014  
698 2014-01-04             4      April   2014  
699 2014-01-05             5        May   2014  

[700 rows x 16 columns]

In [53]:
# Membuat tabel products yang berisi isi list dan menyimpannya ke variabel

products = pd.DataFrame(df['Product'].unique())
products

0
0  Carretera
1    Montana
2      Paseo
3       Velo
4        VTT
5   Amarilla

In [54]:
# Mengganti nama isi product menjadi index

i = 0

# Melakukan proses looping untuk mengakses seluruh baris
for mainName in df['Product']:
    for listName in products.index:
        if mainName == products.iloc[listName][0]:
            df.loc[i, 'Product'] = listName + 1
    i += 1
            
df

Segment Country Product Discount Band  Units Sold  Manufacturing Price  \
0         1       1       1          None      1618.5                  3.0   
1         1       2       1          None      1321.0                  3.0   
2         2       3       1          None      2178.0                  3.0   
3         2       2       1          None       888.0                  3.0   
4         2       4       1          None      2470.0                  3.0   
..      ...     ...     ...           ...         ...                  ...   
695       5       3       6          High      2475.0                260.0   
696       5       4       6          High       546.0                260.0   
697       1       4       2          High      1368.0                  5.0   
698       1       1       3          High       723.0                 10.0   
699       3       5       5          High      1806.0                250.0   

     Sale Price  Gross Sales  Discounts      Sales      COGS    Profit  \
0          20.0      32370.0       0.00   32370.00   16185.0  16185.00   
1          20.0      26420.0       0.00   26420.00   13210.0  13210.00   
2          15.0      32670.0       0.00   32670.00   21780.0  10890.00   
3          15.0      13320.0       0.00   13320.00    8880.0   4440.00   
4          15.0      37050.0       0.00   37050.00   24700.0  12350.00   
..          ...          ...        ...        ...       ...       ...   
695       300.0     742500.0  111375.00  631125.00  618750.0  12375.00   
696       300.0     163800.0   24570.00  139230.00  136500.0   2730.00   
697         7.0       9576.0    1436.40    8139.60    6840.0   1299.60   
698         7.0       5061.0     759.15    4301.85    3615.0    686.85   
699        12.0      21672.0    3250.80   18421.20    5418.0  13003.20   

          Date  Month Number  Month Name  Year  
0   2014-01-01             1    January   2014  
1   2014-01-01             1    January   2014  
2   2014-01-06             6       June   2014  
3   2014-01-06             6       June   2014  
4   2014-01-06             6       June   2014  
..         ...           ...         ...   ...  
695 2014-01-03             3      March   2014  
696 2014-01-10            10    October   2014  
697 2014-01-02             2   February   2014  
698 2014-01-04             4      April   2014  
699 2014-01-05             5        May   2014  

[700 rows x 16 columns]

In [55]:
# Membuat tabel discount bands yang berisi isi list dan menyimpannya ke variabel

discountBands = pd.DataFrame(df['Discount Band'].unique())
discountBands

0
0    None
1     Low
2  Medium
3    High

In [56]:
# Mengganti nama isi discount band menjadi index

i = 0

# Melakukan proses looping untuk mengakses seluruh baris
for mainName in df['Discount Band']:
    for listName in discountBands.index:
        if mainName == discountBands.iloc[listName][0]:
            df.loc[i, 'Discount Band'] = listName + 1
    i += 1

df

Segment Country Product Discount Band  Units Sold  Manufacturing Price  \
0         1       1       1             1      1618.5                  3.0   
1         1       2       1             1      1321.0                  3.0   
2         2       3       1             1      2178.0                  3.0   
3         2       2       1             1       888.0                  3.0   
4         2       4       1             1      2470.0                  3.0   
..      ...     ...     ...           ...         ...                  ...   
695       5       3       6             4      2475.0                260.0   
696       5       4       6             4       546.0                260.0   
697       1       4       2             4      1368.0                  5.0   
698       1       1       3             4       723.0                 10.0   
699       3       5       5             4      1806.0                250.0   

     Sale Price  Gross Sales  Discounts      Sales      COGS    Profit  \
0          20.0      32370.0       0.00   32370.00   16185.0  16185.00   
1          20.0      26420.0       0.00   26420.00   13210.0  13210.00   
2          15.0      32670.0       0.00   32670.00   21780.0  10890.00   
3          15.0      13320.0       0.00   13320.00    8880.0   4440.00   
4          15.0      37050.0       0.00   37050.00   24700.0  12350.00   
..          ...          ...        ...        ...       ...       ...   
695       300.0     742500.0  111375.00  631125.00  618750.0  12375.00   
696       300.0     163800.0   24570.00  139230.00  136500.0   2730.00   
697         7.0       9576.0    1436.40    8139.60    6840.0   1299.60   
698         7.0       5061.0     759.15    4301.85    3615.0    686.85   
699        12.0      21672.0    3250.80   18421.20    5418.0  13003.20   

          Date  Month Number  Month Name  Year  
0   2014-01-01             1    January   2014  
1   2014-01-01             1    January   2014  
2   2014-01-06             6       June   2014  
3   2014-01-06             6       June   2014  
4   2014-01-06             6       June   2014  
..         ...           ...         ...   ...  
695 2014-01-03             3      March   2014  
696 2014-01-10            10    October   2014  
697 2014-01-02             2   February   2014  
698 2014-01-04             4      April   2014  
699 2014-01-05             5        May   2014  

[700 rows x 16 columns]

Pada tahap ini, tabel-tabel baru dibuat, yaitu tabel segments, countries, products, dan discount bands yang berisi list nama dan indeksnya. Indeks ditambahkan dengan angka 1 karena di Visual Studio Code, indeks dimulai dari 0 sementara di PostgreSQL, indeks dimulai dari angka 1. Karena data ini akan diolah ke PostgreSQL selanjutnya, maka indeksnya ditambahkan 1.

Setelah membuat tabel-tabel tersebut, nama list dari tabel utama juga diubah menjadi indeks yang mewakili masing-masing nama.

In [57]:
# Mengubah tipe data dari object menjadi integer

df['Segment'] = df['Segment'].astype(int)
df['Country'] = df['Country'].astype(int)
df['Product'] = df['Product'].astype(int)
df['Discount Band'] = df['Discount Band'].astype(int)

In [58]:
# Menampilkan tampilan akhir dari main table

df

Segment  Country  Product  Discount Band  Units Sold  \
0          1        1        1              1      1618.5   
1          1        2        1              1      1321.0   
2          2        3        1              1      2178.0   
3          2        2        1              1       888.0   
4          2        4        1              1      2470.0   
..       ...      ...      ...            ...         ...   
695        5        3        6              4      2475.0   
696        5        4        6              4       546.0   
697        1        4        2              4      1368.0   
698        1        1        3              4       723.0   
699        3        5        5              4      1806.0   

     Manufacturing Price  Sale Price  Gross Sales  Discounts      Sales  \
0                    3.0        20.0      32370.0       0.00   32370.00   
1                    3.0        20.0      26420.0       0.00   26420.00   
2                    3.0        15.0      32670.0       0.00   32670.00   
3                    3.0        15.0      13320.0       0.00   13320.00   
4                    3.0        15.0      37050.0       0.00   37050.00   
..                   ...         ...          ...        ...        ...   
695                260.0       300.0     742500.0  111375.00  631125.00   
696                260.0       300.0     163800.0   24570.00  139230.00   
697                  5.0         7.0       9576.0    1436.40    8139.60   
698                 10.0         7.0       5061.0     759.15    4301.85   
699                250.0        12.0      21672.0    3250.80   18421.20   

         COGS    Profit       Date  Month Number  Month Name  Year  
0     16185.0  16185.00 2014-01-01             1    January   2014  
1     13210.0  13210.00 2014-01-01             1    January   2014  
2     21780.0  10890.00 2014-01-06             6       June   2014  
3      8880.0   4440.00 2014-01-06             6       June   2014  
4     24700.0  12350.00 2014-01-06             6       June   2014  
..        ...       ...        ...           ...         ...   ...  
695  618750.0  12375.00 2014-01-03             3      March   2014  
696  136500.0   2730.00 2014-01-10            10    October   2014  
697    6840.0   1299.60 2014-01-02             2   February   2014  
698    3615.0    686.85 2014-01-04             4      April   2014  
699    5418.0  13003.20 2014-01-05             5        May   2014  

[700 rows x 16 columns]

Dataframe di atas merupakan main table dari data yang telah dinormalisasi. Dapat lihat bahwa kolom `Segment`, `Country`, `Product`, dan `Discount Band` telah diganti menjadi angka yang merepresentasikan setiap nama berdasarkan tabel masing-masing.

##### d. Kemudian simpan semua tabel yang dihasilkan ke dalam file csv. Berikan nama file sesuai dengan nama tabel yang akan digunakan pada database. Jangan lupa index dataframe tidak perlu diikutsertakan dalam file karena skema tabel akan menerapkan auto incremental untuk pengisian primay key

In [59]:
# Menyimpan kelima table menjadi format CSV tanpa menyertakan index

df.to_csv('mainTable.csv', index=False)
segments.to_csv('segments.csv', index=False)
countries.to_csv('countries.csv', index=False)
products.to_csv('products.csv', index=False)
discountBands.to_csv('discountBands.csv', index=False)

Setelah melakukan eksplorasi, *cleaning*, dan normalisasi data, kelima dataframe sudah dapat disimpan ke dalam format CSV dengan metode `to_csv` untuk dilanjutkan ke tahap selanjutnya.

#### **Kesimpulan**

Kesimpulan dari proses ini adalah setelah melakukan eksplorasi data, dataframe yang diterima masih mengandung banyak data yang tidak bersih dan tidak sesuai. Oleh karena itu, dilakukan proses *cleaning* data dengan mengganti nama kolom yang tidak jelas, memperbarui nilai-nilai dalam kolom agar sesuai dengan format yang diinginkan, menghapus karakter yang tidak perlu, dan menangani nilai-nilai yang hilang *(missing values)*. Setelah proses *cleaning* selesai, dilakukan proses normalisasi hingga mencapai bentuk normalisasi ketiga (3NF) untuk memastikan struktur data yang lebih baik dan meminimalkan redundansi.

#### **Hasil SQL**

##### 1. a. Berikan tabel yang berisikan informasi total profit yang didapatkan di tiap jenis segmentasi. Jangan ambil data yang tidak diskon.

```sql
SELECT a.name as "Segment",
	   ROUND(SUM(b.profit)::numeric, 2) as "Total Profit"
FROM segments as a
JOIN mainTable as b
ON a.id = b.segment
WHERE b.discounts > 0
GROUP BY 1
ORDER BY 2 DESC;
```

In [62]:
tabelQuery1 = pd.read_csv("tabelQuery1.csv")
tabelQuery1

Segment  Total Profit
0        Government   10270950.18
1    Small Business    3848268.50
2  Channel Partners    1150321.14
3        Enterprise     879096.88
4         Midmarket     563253.09

Berdasarkan hasil di atas, dapat disimpulkan bahwa total profit **tertinggi** diambil oleh segment ***Government*** dengan nilai sebesar 10270950.18, diikuti oleh ***Small Business*** dengan nilai sebesar 3848268.50, lalu ***Channel Partners*** dengan nilai sebesar 1150321.14, lalu ***Enterprise*** dengan nilai sebesar 879096.88, dan terakhir adalah ***Midmarket*** dengan nilai sebesar 563253.09.

##### 1. b. Berikan tabel yang berisikan informasi rangkuman statistik yang memuat nilai rata-rata, min, dan max (dijadikan dalam kolom yang berbeda) dari Sales masing-masing negara.

```sql
SELECT a.name as "Country",
	   ROUND(AVG(b.sales)::numeric, 2) as "Average Sales",
	   MIN(b.sales) as "Minimum Sales",
	   MAX(b.sales) as "Maximum Sales"
FROM countries as a
JOIN mainTable as b
ON a.id = b.country
GROUP BY 1;
```

In [63]:
tabelQuery2 = pd.read_csv("tabelQuery2.csv")
tabelQuery2

Country  Average Sales  Minimum Sales  Maximum Sales
0                    France      173958.37        1763.86       962500.0
1                   Germany      167895.29        1685.60      1017338.0
2                    Canada      177768.96        1655.08      1038082.5
3  United States of America      178784.50        1730.54      1159200.0
4                    Mexico      149638.23        2508.66       848172.5

Berdasarkan hasil di atas, terdapat 3 nilai statistik yang dihitung, yaitu rata-rata, nilai minimum, dan nilai maksimum dari Sales masing-masing negara.
- **Rata-rata** sales **tertinggi** diraih oleh negara **United States of America** dengan nilai sebesar 178784.50 dan **terendah** diraih oleh negara **Mexico** dengan nilai sebesar 149638.23. 
- **Nilai minimum**  sales **terbesar** diraih oleh negara **Mexico** dengan nilai sebesar 2508.66 dan nilai **minimum terkecil** diraih oleh negara **Canada** dengan nilai sebesar 1655.08.
- **Nilai maximum** sales **terbesar** diraih oleh negara **United States of America** dengan nilai sebesar 1159200.0 dan nilai **maximum terkecil** diraih oleh negara **Mexico** dengan nilai sebesar 8488172.5.